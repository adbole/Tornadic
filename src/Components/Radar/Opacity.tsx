import React from "react";

import { useBooleanState } from "Hooks";

import { CircleSlashes } from "../../svgs/radar";


export default function Opacity({
    value,
    setOpacity,
}: {
    value: number;
    setOpacity: (x: number) => void;
}) {
    const [hover, setHoverTrue, setHoverFalse] = useBooleanState(false);

    //Fallback for touch devices
    React.useEffect(() => {
        document.body.addEventListener("click", setHoverFalse);
        return () => document.body.removeEventListener("click", setHoverFalse);
    }, [setHoverFalse]);

    return (
        <div
            className="leaflet-custom-control leaflet-control"
            onClick={(e: React.MouseEvent<HTMLDivElement>) => e.stopPropagation()}
            onMouseEnter={setHoverTrue}
            onMouseLeave={setHoverFalse}
        >
            {!hover && (
                <div className="leaflet-control-toggle">
                    <CircleSlashes />
                </div>
            )}
            {hover && (
                <div id="opacity">
                    <p>Opacity: {(value * 100).toFixed(0)}</p>
                    <input
                        type="range"
                        min={0}
                        max={100}
                        value={value * 100}
                        step={1}
                        onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
                            setOpacity(e.currentTarget.valueAsNumber / 100)
                        }
                    />
                </div>
            )}
        </div>
    );
}
