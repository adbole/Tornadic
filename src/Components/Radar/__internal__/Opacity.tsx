import React from "react";
import { useMap } from "react-leaflet";

import { Range } from "Components/Input";


export default function Opacity({
    defaultOpacity,
    targetPane,
}: {
    defaultOpacity: number;
    targetPane: string;
}) {
    const map = useMap();

    const [value, setOpacity] = React.useState(defaultOpacity);

    React.useEffect(() => {
        const pane = map.getPane(targetPane);
        if (!pane) {
            const observer = new MutationObserver((_, observer) => {
                const pane = map.getPane(targetPane);

                if (!pane) return;

                pane.style.opacity = value.toString();

                observer.disconnect();
            });

            observer.observe(map.getContainer(), { childList: true, subtree: true });

            return;
        }

        pane.style.opacity = value.toString();
    }, [map, targetPane, value]);

    return (
        <label>
            <p style={{ textAlign: "center" }}>Opacity: {(value * 100).toFixed(0)}</p>
            <Range
                min={0}
                max={100}
                value={value * 100}
                step={1}
                onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
                    setOpacity(e.currentTarget.valueAsNumber / 100)
                }
            />
        </label>
    );
}
