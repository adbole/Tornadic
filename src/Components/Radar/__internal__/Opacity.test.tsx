import { MapContainer, useMap } from "react-leaflet";
import { act, fireEvent, render, renderHook, screen } from "@testing-library/react";

import Opacity from "./Opacity";


function WrapperInteraction({ children }: { children: React.ReactNode }) {
    return (
        <MapContainer>
            {children}
            <Opacity defaultOpacity={0.8} targetPane="tilePane" />
        </MapContainer>
    );
}

function WrapperWait({ children }: { children: React.ReactNode }) {
    return (
        <MapContainer>
            <Opacity defaultOpacity={0.8} targetPane="radar" />
            {children}
        </MapContainer>
    );
}

test("Matches snapshot", () => {
    const { container } = render(
        <MapContainer>
            <Opacity defaultOpacity={0.8} targetPane="tilePane" />
        </MapContainer>
    );

    expect(container).toMatchSnapshot();
});

test("Shows a slider and the current opacity", () => {
    render(
        <MapContainer>
            <Opacity defaultOpacity={0.37} targetPane="tilePane" />
        </MapContainer>
    );

    expect.soft(screen.queryByText("Opacity: 37")).toBeInTheDocument();
    expect.soft(screen.queryByRole("slider")).toBeInTheDocument();
});

test("Sets the target pane's opacity when slider is changed", () => {
    const {
        result: { current: map },
    } = renderHook(useMap, { wrapper: WrapperInteraction });

    const slider = screen.queryByRole("slider") as HTMLInputElement;

    act(() => {
        fireEvent.change(slider, { target: { value: 50 } });
    });

    expect.soft(map.getPane("tilePane")).toHaveStyle({ opacity: "0.5" });
});

test("If the target pane doesn't exist, then a mutation observer waits for its creation", () => {
    vi.spyOn(window.MutationObserver.prototype, "observe");

    const {
        result: { current: map },
    } = renderHook(useMap, { wrapper: WrapperWait });

    expect.soft(map.getPane("radar")).not.toBeTruthy();
    map.createPane("radar");

    const slider = screen.queryByRole("slider") as HTMLInputElement;

    act(() => {
        fireEvent.change(slider, { target: { value: 50 } });
    });

    expect.soft(window.MutationObserver.prototype.observe).toHaveBeenCalledOnce();
    expect.soft(map.getPane("radar")).toHaveStyle({ opacity: "0.5" });
});
