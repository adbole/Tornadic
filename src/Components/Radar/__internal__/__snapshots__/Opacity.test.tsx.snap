// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Matches snapshot 1`] = `
.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background: transparent;
  cursor: pointer;
  width: 100%;
}

.emotion-0::-webkit-slider-runnable-track {
  background-color: rgba(136, 136, 136, 0.5);
  border-radius: var(--border-radius);
}

.emotion-0::-webkit-slider-thumb {
  -webkit-appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  height: 10px;
  width: 10px;
  border-radius: var(--border-radius);
  background-color: #6498fa;
}

<div>
  <div
    class="leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom"
    style="position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
        style="opacity: 0.8;"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-disabled="false"
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            <span
              aria-hidden="true"
            >
              +
            </span>
          </a>
          <a
            aria-disabled="false"
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            <span
              aria-hidden="true"
            >
              âˆ’
            </span>
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JavaScript library for interactive maps"
          >
            Leaflet
          </a>
        </div>
      </div>
    </div>
    <div>
      <p
        style="text-align: center;"
      >
        Opacity: 
        80
      </p>
      <input
        class="emotion-0 emotion-1"
        max="100"
        min="0"
        step="1"
        title="Opacity"
        type="range"
        value="80"
      />
    </div>
  </div>
</div>
`;
